{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy  as np\n",
    "from   scipy  import stats\n",
    "\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.preprocessing   import StandardScaler\n",
    "from sklearn.decomposition   import PCA\n",
    "from sklearn.metrics         import accuracy_score\n",
    "from sklearn.neighbors       import KNeighborsClassifier\n",
    "from sklearn.svm             import SVC \n",
    "from sklearn.ensemble        import RandomForestClassifier \n",
    "from sklearn.naive_bayes     import GaussianNB\n",
    "\n",
    "pd.options.mode.chained_assignment = None\n",
    "\n",
    "LUMINAL_A = \"Luminal A    \"\n",
    "LUMINAL_B = \"Luminal B    \""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Dataset exploration"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Label</th>\n",
       "      <th>ENSG00000242268.2</th>\n",
       "      <th>ENSG00000270112.3</th>\n",
       "      <th>ENSG00000167578.15</th>\n",
       "      <th>ENSG00000273842.1</th>\n",
       "      <th>ENSG00000078237.5</th>\n",
       "      <th>ENSG00000146083.10</th>\n",
       "      <th>ENSG00000225275.4</th>\n",
       "      <th>ENSG00000158486.12</th>\n",
       "      <th>ENSG00000198242.12</th>\n",
       "      <th>...</th>\n",
       "      <th>ENSG00000088356.5</th>\n",
       "      <th>ENSG00000176752.6</th>\n",
       "      <th>ENSG00000223082.1</th>\n",
       "      <th>ENSG00000237714.1</th>\n",
       "      <th>ENSG00000200959.1</th>\n",
       "      <th>ENSG00000270971.2</th>\n",
       "      <th>ENSG00000267313.5</th>\n",
       "      <th>ENSG00000151632.15</th>\n",
       "      <th>ENSG00000269107.1</th>\n",
       "      <th>ENSG00000268889.1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Luminal A</td>\n",
       "      <td>2356.253792</td>\n",
       "      <td>26808.891103</td>\n",
       "      <td>57790.161586</td>\n",
       "      <td>0.0</td>\n",
       "      <td>71389.111749</td>\n",
       "      <td>381288.489078</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1677.413295</td>\n",
       "      <td>1.969141e+06</td>\n",
       "      <td>...</td>\n",
       "      <td>113036.499486</td>\n",
       "      <td>0.0</td>\n",
       "      <td>20911.752405</td>\n",
       "      <td>13957.784680</td>\n",
       "      <td>1.047700e+06</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1937.768563</td>\n",
       "      <td>10731.974977</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Luminal A</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>231.960840</td>\n",
       "      <td>115769.964478</td>\n",
       "      <td>0.0</td>\n",
       "      <td>77938.573803</td>\n",
       "      <td>238017.846096</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.865670e+06</td>\n",
       "      <td>...</td>\n",
       "      <td>850405.301444</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2122.068451</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5499.352680</td>\n",
       "      <td>2479.070667</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Luminal A</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>315.873536</td>\n",
       "      <td>44954.933833</td>\n",
       "      <td>0.0</td>\n",
       "      <td>58546.997851</td>\n",
       "      <td>249302.406604</td>\n",
       "      <td>0.0</td>\n",
       "      <td>270.107768</td>\n",
       "      <td>2.519212e+06</td>\n",
       "      <td>...</td>\n",
       "      <td>301948.579552</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>481.622432</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>312.031830</td>\n",
       "      <td>4822.690978</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Luminal A</td>\n",
       "      <td>1074.333108</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>86991.783442</td>\n",
       "      <td>0.0</td>\n",
       "      <td>36082.381881</td>\n",
       "      <td>176274.146728</td>\n",
       "      <td>0.0</td>\n",
       "      <td>101.975469</td>\n",
       "      <td>2.855010e+06</td>\n",
       "      <td>...</td>\n",
       "      <td>293581.147891</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>12631.387770</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Luminal A</td>\n",
       "      <td>1395.887715</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>65199.337535</td>\n",
       "      <td>0.0</td>\n",
       "      <td>91469.201926</td>\n",
       "      <td>263704.398926</td>\n",
       "      <td>0.0</td>\n",
       "      <td>66.248682</td>\n",
       "      <td>4.235429e+06</td>\n",
       "      <td>...</td>\n",
       "      <td>350449.557585</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1181.263740</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1626.418477</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 1023 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           Label  ENSG00000242268.2  ENSG00000270112.3  ENSG00000167578.15  \\\n",
       "0  Luminal A            2356.253792       26808.891103        57790.161586   \n",
       "1  Luminal A               0.000000         231.960840       115769.964478   \n",
       "2  Luminal A               0.000000         315.873536        44954.933833   \n",
       "3  Luminal A            1074.333108           0.000000        86991.783442   \n",
       "4  Luminal A            1395.887715           0.000000        65199.337535   \n",
       "\n",
       "   ENSG00000273842.1  ENSG00000078237.5  ENSG00000146083.10  \\\n",
       "0                0.0       71389.111749       381288.489078   \n",
       "1                0.0       77938.573803       238017.846096   \n",
       "2                0.0       58546.997851       249302.406604   \n",
       "3                0.0       36082.381881       176274.146728   \n",
       "4                0.0       91469.201926       263704.398926   \n",
       "\n",
       "   ENSG00000225275.4  ENSG00000158486.12  ENSG00000198242.12  ...  \\\n",
       "0                0.0         1677.413295        1.969141e+06  ...   \n",
       "1                0.0            0.000000        4.865670e+06  ...   \n",
       "2                0.0          270.107768        2.519212e+06  ...   \n",
       "3                0.0          101.975469        2.855010e+06  ...   \n",
       "4                0.0           66.248682        4.235429e+06  ...   \n",
       "\n",
       "   ENSG00000088356.5  ENSG00000176752.6  ENSG00000223082.1  ENSG00000237714.1  \\\n",
       "0      113036.499486                0.0       20911.752405       13957.784680   \n",
       "1      850405.301444                0.0           0.000000        2122.068451   \n",
       "2      301948.579552                0.0           0.000000         481.622432   \n",
       "3      293581.147891                0.0           0.000000           0.000000   \n",
       "4      350449.557585                0.0           0.000000        1181.263740   \n",
       "\n",
       "   ENSG00000200959.1  ENSG00000270971.2  ENSG00000267313.5  \\\n",
       "0       1.047700e+06                0.0        1937.768563   \n",
       "1       0.000000e+00                0.0        5499.352680   \n",
       "2       0.000000e+00                0.0         312.031830   \n",
       "3       0.000000e+00                0.0           0.000000   \n",
       "4       0.000000e+00                0.0           0.000000   \n",
       "\n",
       "   ENSG00000151632.15  ENSG00000269107.1  ENSG00000268889.1  \n",
       "0        10731.974977                0.0                0.0  \n",
       "1         2479.070667                0.0                0.0  \n",
       "2         4822.690978                0.0                0.0  \n",
       "3        12631.387770                0.0                0.0  \n",
       "4         1626.418477                0.0                0.0  \n",
       "\n",
       "[5 rows x 1023 columns]"
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "luminal_A_B_df = pd.read_csv(\"dataset_LUMINAL_A_B.csv\").rename(columns={'l': 'Label'})\n",
    "\n",
    "genes = list(luminal_A_B_df.columns)\n",
    "genes.remove(\"Label\")\n",
    "\n",
    "luminal_A_B_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ENSG00000242268.2</th>\n",
       "      <th>ENSG00000270112.3</th>\n",
       "      <th>ENSG00000167578.15</th>\n",
       "      <th>ENSG00000273842.1</th>\n",
       "      <th>ENSG00000078237.5</th>\n",
       "      <th>ENSG00000146083.10</th>\n",
       "      <th>ENSG00000225275.4</th>\n",
       "      <th>ENSG00000158486.12</th>\n",
       "      <th>ENSG00000198242.12</th>\n",
       "      <th>ENSG00000259883.1</th>\n",
       "      <th>...</th>\n",
       "      <th>ENSG00000088356.5</th>\n",
       "      <th>ENSG00000176752.6</th>\n",
       "      <th>ENSG00000223082.1</th>\n",
       "      <th>ENSG00000237714.1</th>\n",
       "      <th>ENSG00000200959.1</th>\n",
       "      <th>ENSG00000270971.2</th>\n",
       "      <th>ENSG00000267313.5</th>\n",
       "      <th>ENSG00000151632.15</th>\n",
       "      <th>ENSG00000269107.1</th>\n",
       "      <th>ENSG00000268889.1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2356.253792</td>\n",
       "      <td>26808.891103</td>\n",
       "      <td>57790.161586</td>\n",
       "      <td>0.0</td>\n",
       "      <td>71389.111749</td>\n",
       "      <td>381288.489078</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1677.413295</td>\n",
       "      <td>1.969141e+06</td>\n",
       "      <td>86793.265493</td>\n",
       "      <td>...</td>\n",
       "      <td>113036.499486</td>\n",
       "      <td>0.0</td>\n",
       "      <td>20911.752405</td>\n",
       "      <td>13957.784680</td>\n",
       "      <td>1.047700e+06</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1937.768563</td>\n",
       "      <td>10731.974977</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>231.960840</td>\n",
       "      <td>115769.964478</td>\n",
       "      <td>0.0</td>\n",
       "      <td>77938.573803</td>\n",
       "      <td>238017.846096</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.865670e+06</td>\n",
       "      <td>769.742944</td>\n",
       "      <td>...</td>\n",
       "      <td>850405.301444</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2122.068451</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5499.352680</td>\n",
       "      <td>2479.070667</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>315.873536</td>\n",
       "      <td>44954.933833</td>\n",
       "      <td>0.0</td>\n",
       "      <td>58546.997851</td>\n",
       "      <td>249302.406604</td>\n",
       "      <td>0.0</td>\n",
       "      <td>270.107768</td>\n",
       "      <td>2.519212e+06</td>\n",
       "      <td>2620.500780</td>\n",
       "      <td>...</td>\n",
       "      <td>301948.579552</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>481.622432</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>312.031830</td>\n",
       "      <td>4822.690978</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1074.333108</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>86991.783442</td>\n",
       "      <td>0.0</td>\n",
       "      <td>36082.381881</td>\n",
       "      <td>176274.146728</td>\n",
       "      <td>0.0</td>\n",
       "      <td>101.975469</td>\n",
       "      <td>2.855010e+06</td>\n",
       "      <td>9893.339903</td>\n",
       "      <td>...</td>\n",
       "      <td>293581.147891</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>12631.387770</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1395.887715</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>65199.337535</td>\n",
       "      <td>0.0</td>\n",
       "      <td>91469.201926</td>\n",
       "      <td>263704.398926</td>\n",
       "      <td>0.0</td>\n",
       "      <td>66.248682</td>\n",
       "      <td>4.235429e+06</td>\n",
       "      <td>2570.895661</td>\n",
       "      <td>...</td>\n",
       "      <td>350449.557585</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1181.263740</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1626.418477</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 1022 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   ENSG00000242268.2  ENSG00000270112.3  ENSG00000167578.15  \\\n",
       "0        2356.253792       26808.891103        57790.161586   \n",
       "1           0.000000         231.960840       115769.964478   \n",
       "2           0.000000         315.873536        44954.933833   \n",
       "3        1074.333108           0.000000        86991.783442   \n",
       "4        1395.887715           0.000000        65199.337535   \n",
       "\n",
       "   ENSG00000273842.1  ENSG00000078237.5  ENSG00000146083.10  \\\n",
       "0                0.0       71389.111749       381288.489078   \n",
       "1                0.0       77938.573803       238017.846096   \n",
       "2                0.0       58546.997851       249302.406604   \n",
       "3                0.0       36082.381881       176274.146728   \n",
       "4                0.0       91469.201926       263704.398926   \n",
       "\n",
       "   ENSG00000225275.4  ENSG00000158486.12  ENSG00000198242.12  \\\n",
       "0                0.0         1677.413295        1.969141e+06   \n",
       "1                0.0            0.000000        4.865670e+06   \n",
       "2                0.0          270.107768        2.519212e+06   \n",
       "3                0.0          101.975469        2.855010e+06   \n",
       "4                0.0           66.248682        4.235429e+06   \n",
       "\n",
       "   ENSG00000259883.1  ...  ENSG00000088356.5  ENSG00000176752.6  \\\n",
       "0       86793.265493  ...      113036.499486                0.0   \n",
       "1         769.742944  ...      850405.301444                0.0   \n",
       "2        2620.500780  ...      301948.579552                0.0   \n",
       "3        9893.339903  ...      293581.147891                0.0   \n",
       "4        2570.895661  ...      350449.557585                0.0   \n",
       "\n",
       "   ENSG00000223082.1  ENSG00000237714.1  ENSG00000200959.1  ENSG00000270971.2  \\\n",
       "0       20911.752405       13957.784680       1.047700e+06                0.0   \n",
       "1           0.000000        2122.068451       0.000000e+00                0.0   \n",
       "2           0.000000         481.622432       0.000000e+00                0.0   \n",
       "3           0.000000           0.000000       0.000000e+00                0.0   \n",
       "4           0.000000        1181.263740       0.000000e+00                0.0   \n",
       "\n",
       "   ENSG00000267313.5  ENSG00000151632.15  ENSG00000269107.1  ENSG00000268889.1  \n",
       "0        1937.768563        10731.974977                0.0                0.0  \n",
       "1        5499.352680         2479.070667                0.0                0.0  \n",
       "2         312.031830         4822.690978                0.0                0.0  \n",
       "3           0.000000        12631.387770                0.0                0.0  \n",
       "4           0.000000         1626.418477                0.0                0.0  \n",
       "\n",
       "[5 rows x 1022 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "luminal_A_df = luminal_A_B_df[luminal_A_B_df.Label == LUMINAL_A].drop(\"Label\", axis=1)\n",
    "luminal_A_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ENSG00000242268.2</th>\n",
       "      <th>ENSG00000270112.3</th>\n",
       "      <th>ENSG00000167578.15</th>\n",
       "      <th>ENSG00000273842.1</th>\n",
       "      <th>ENSG00000078237.5</th>\n",
       "      <th>ENSG00000146083.10</th>\n",
       "      <th>ENSG00000225275.4</th>\n",
       "      <th>ENSG00000158486.12</th>\n",
       "      <th>ENSG00000198242.12</th>\n",
       "      <th>ENSG00000259883.1</th>\n",
       "      <th>...</th>\n",
       "      <th>ENSG00000088356.5</th>\n",
       "      <th>ENSG00000176752.6</th>\n",
       "      <th>ENSG00000223082.1</th>\n",
       "      <th>ENSG00000237714.1</th>\n",
       "      <th>ENSG00000200959.1</th>\n",
       "      <th>ENSG00000270971.2</th>\n",
       "      <th>ENSG00000267313.5</th>\n",
       "      <th>ENSG00000151632.15</th>\n",
       "      <th>ENSG00000269107.1</th>\n",
       "      <th>ENSG00000268889.1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>472.317473</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>43185.197034</td>\n",
       "      <td>0.0</td>\n",
       "      <td>71328.844164</td>\n",
       "      <td>197593.096861</td>\n",
       "      <td>0.0</td>\n",
       "      <td>381.074325</td>\n",
       "      <td>2.544896e+06</td>\n",
       "      <td>3914.537811</td>\n",
       "      <td>...</td>\n",
       "      <td>168679.746145</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1598.786202</td>\n",
       "      <td>1693.663665</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>26915.683769</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>115939.159122</td>\n",
       "      <td>0.0</td>\n",
       "      <td>97148.626331</td>\n",
       "      <td>197179.900505</td>\n",
       "      <td>0.0</td>\n",
       "      <td>177.495804</td>\n",
       "      <td>6.747133e+06</td>\n",
       "      <td>1722.008578</td>\n",
       "      <td>...</td>\n",
       "      <td>359360.954156</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3319.171535</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>102.522673</td>\n",
       "      <td>396.141612</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1425.046813</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>61.848883</td>\n",
       "      <td>45418.241007</td>\n",
       "      <td>0.0</td>\n",
       "      <td>36746.522566</td>\n",
       "      <td>219188.070354</td>\n",
       "      <td>0.0</td>\n",
       "      <td>359.637198</td>\n",
       "      <td>9.354113e+06</td>\n",
       "      <td>1231.442546</td>\n",
       "      <td>...</td>\n",
       "      <td>119115.871944</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1131.635522</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>472.147895</td>\n",
       "      <td>0.0</td>\n",
       "      <td>566.155160</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>89420.856204</td>\n",
       "      <td>0.0</td>\n",
       "      <td>80340.679862</td>\n",
       "      <td>241397.554846</td>\n",
       "      <td>0.0</td>\n",
       "      <td>59.576577</td>\n",
       "      <td>2.894877e+06</td>\n",
       "      <td>4045.952776</td>\n",
       "      <td>...</td>\n",
       "      <td>288613.033711</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1062.295110</td>\n",
       "      <td>9002.682799</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2393.373228</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>218106.422474</td>\n",
       "      <td>0.0</td>\n",
       "      <td>86026.755295</td>\n",
       "      <td>228346.920871</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4887.091780</td>\n",
       "      <td>2.897671e+06</td>\n",
       "      <td>3818.499101</td>\n",
       "      <td>...</td>\n",
       "      <td>455313.016648</td>\n",
       "      <td>0.0</td>\n",
       "      <td>12266.928361</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2478.167346</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3879.738643</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 1022 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    ENSG00000242268.2  ENSG00000270112.3  ENSG00000167578.15  \\\n",
       "50         472.317473           0.000000        43185.197034   \n",
       "51           0.000000           0.000000       115939.159122   \n",
       "52           0.000000          61.848883        45418.241007   \n",
       "53           0.000000           0.000000        89420.856204   \n",
       "54           0.000000           0.000000       218106.422474   \n",
       "\n",
       "    ENSG00000273842.1  ENSG00000078237.5  ENSG00000146083.10  \\\n",
       "50                0.0       71328.844164       197593.096861   \n",
       "51                0.0       97148.626331       197179.900505   \n",
       "52                0.0       36746.522566       219188.070354   \n",
       "53                0.0       80340.679862       241397.554846   \n",
       "54                0.0       86026.755295       228346.920871   \n",
       "\n",
       "    ENSG00000225275.4  ENSG00000158486.12  ENSG00000198242.12  \\\n",
       "50                0.0          381.074325        2.544896e+06   \n",
       "51                0.0          177.495804        6.747133e+06   \n",
       "52                0.0          359.637198        9.354113e+06   \n",
       "53                0.0           59.576577        2.894877e+06   \n",
       "54                0.0         4887.091780        2.897671e+06   \n",
       "\n",
       "    ENSG00000259883.1  ...  ENSG00000088356.5  ENSG00000176752.6  \\\n",
       "50        3914.537811  ...      168679.746145                0.0   \n",
       "51        1722.008578  ...      359360.954156                0.0   \n",
       "52        1231.442546  ...      119115.871944                0.0   \n",
       "53        4045.952776  ...      288613.033711                0.0   \n",
       "54        3818.499101  ...      455313.016648                0.0   \n",
       "\n",
       "    ENSG00000223082.1  ENSG00000237714.1  ENSG00000200959.1  \\\n",
       "50           0.000000        1598.786202        1693.663665   \n",
       "51        3319.171535           0.000000           0.000000   \n",
       "52           0.000000        1131.635522           0.000000   \n",
       "53           0.000000        1062.295110        9002.682799   \n",
       "54       12266.928361           0.000000        2478.167346   \n",
       "\n",
       "    ENSG00000270971.2  ENSG00000267313.5  ENSG00000151632.15  \\\n",
       "50                0.0           0.000000        26915.683769   \n",
       "51                0.0         102.522673          396.141612   \n",
       "52                0.0           0.000000          472.147895   \n",
       "53                0.0           0.000000         2393.373228   \n",
       "54                0.0           0.000000         3879.738643   \n",
       "\n",
       "    ENSG00000269107.1  ENSG00000268889.1  \n",
       "50                0.0           0.000000  \n",
       "51                0.0        1425.046813  \n",
       "52                0.0         566.155160  \n",
       "53                0.0           0.000000  \n",
       "54                0.0           0.000000  \n",
       "\n",
       "[5 rows x 1022 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "luminal_B_df = luminal_A_B_df[luminal_A_B_df.Label == LUMINAL_B].drop(\"Label\", axis=1)\n",
    "luminal_B_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Assignment 1: T-test to perform differential gene expression analysis\n",
    "\n",
    "Perform a differential gene expression analysis on dataset_LUMINAL_A_B.csv in order to find which genes are able to distinguish between breast cancer Luminal A subtype and breast cancer Luminal B subtype.<br>\n",
    "* Calculate t-value and p-value for each gene and select only genes for which p-value < adjusted Bonferroni p-value. <br>\n",
    "* Finally, convert differentially expressed genes from ENSEMBL notation to the common name (e.g. ENSG00000268889 is AC008750). <br>\n",
    "* Which genes are differentially expressed between the two populations? <br>\n",
    "* Create now a reduced dataset (from now on referred to as reduced_dataset.csv) made up of all samples with only differentially expressed genes (use common name notation)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ENSG00000174013',\n",
       " 'ENSG00000109805',\n",
       " 'ENSG00000127329',\n",
       " 'ENSG00000269113',\n",
       " 'ENSG00000049239',\n",
       " 'ENSG00000267206',\n",
       " 'ENSG00000175643',\n",
       " 'ENSG00000085999',\n",
       " 'ENSG00000167178']"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ALPHA              = 0.05\n",
    "BONFERRONI_P_VALUE = ALPHA / len(genes)\n",
    "\n",
    "differentially_expressed_genes = perform_ttest(luminal_A_df, luminal_B_df, genes)\n",
    "\n",
    "list(differentially_expressed_genes.keys())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneId</th>\n",
       "      <th>GeneName</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000223972</td>\n",
       "      <td>DDX11L1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000227232</td>\n",
       "      <td>WASH7P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000278267</td>\n",
       "      <td>MIR6859-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000243485</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000284332</td>\n",
       "      <td>MIR1302-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58730</th>\n",
       "      <td>ENSG00000271254</td>\n",
       "      <td>AC240274.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58731</th>\n",
       "      <td>ENSG00000275405</td>\n",
       "      <td>RF00003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58732</th>\n",
       "      <td>ENSG00000275987</td>\n",
       "      <td>RF00003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58733</th>\n",
       "      <td>ENSG00000277475</td>\n",
       "      <td>AC213203.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58734</th>\n",
       "      <td>ENSG00000268674</td>\n",
       "      <td>AC213203.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>58735 rows Ã— 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                GeneId     GeneName\n",
       "0      ENSG00000223972      DDX11L1\n",
       "1      ENSG00000227232       WASH7P\n",
       "2      ENSG00000278267    MIR6859-1\n",
       "3      ENSG00000243485  MIR1302-2HG\n",
       "4      ENSG00000284332    MIR1302-2\n",
       "...                ...          ...\n",
       "58730  ENSG00000271254   AC240274.1\n",
       "58731  ENSG00000275405      RF00003\n",
       "58732  ENSG00000275987      RF00003\n",
       "58733  ENSG00000277475   AC213203.2\n",
       "58734  ENSG00000268674   AC213203.1\n",
       "\n",
       "[58735 rows x 2 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "GTF_FILE = \"Homo_sapiens.GRCh38.95.gtf\"\n",
    "\n",
    "file   = parse_gtf(GTF_FILE, prune = True)\n",
    "gtf_df = pd.DataFrame(file, columns={\"GeneId\", \"GeneName\"})\n",
    "\n",
    "gtf_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneId</th>\n",
       "      <th>GeneName</th>\n",
       "      <th>FullGeneId</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>313</th>\n",
       "      <td>ENSG00000049239</td>\n",
       "      <td>H6PD</td>\n",
       "      <td>ENSG00000049239.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1480</th>\n",
       "      <td>ENSG00000085999</td>\n",
       "      <td>RAD54L</td>\n",
       "      <td>ENSG00000085999.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1528</th>\n",
       "      <td>ENSG00000269113</td>\n",
       "      <td>TRABD2B</td>\n",
       "      <td>ENSG00000269113.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12320</th>\n",
       "      <td>ENSG00000174013</td>\n",
       "      <td>FBXO45</td>\n",
       "      <td>ENSG00000174013.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12744</th>\n",
       "      <td>ENSG00000109805</td>\n",
       "      <td>NCAPG</td>\n",
       "      <td>ENSG00000109805.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30524</th>\n",
       "      <td>ENSG00000267206</td>\n",
       "      <td>LCN6</td>\n",
       "      <td>ENSG00000267206.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37901</th>\n",
       "      <td>ENSG00000127329</td>\n",
       "      <td>PTPRB</td>\n",
       "      <td>ENSG00000127329.13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44125</th>\n",
       "      <td>ENSG00000167178</td>\n",
       "      <td>ISLR2</td>\n",
       "      <td>ENSG00000167178.14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45354</th>\n",
       "      <td>ENSG00000175643</td>\n",
       "      <td>RMI2</td>\n",
       "      <td>ENSG00000175643.8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                GeneId GeneName          FullGeneId\n",
       "313    ENSG00000049239     H6PD  ENSG00000049239.11\n",
       "1480   ENSG00000085999   RAD54L  ENSG00000085999.10\n",
       "1528   ENSG00000269113  TRABD2B   ENSG00000269113.3\n",
       "12320  ENSG00000174013   FBXO45   ENSG00000174013.7\n",
       "12744  ENSG00000109805    NCAPG   ENSG00000109805.8\n",
       "30524  ENSG00000267206     LCN6   ENSG00000267206.4\n",
       "37901  ENSG00000127329    PTPRB  ENSG00000127329.13\n",
       "44125  ENSG00000167178    ISLR2  ENSG00000167178.14\n",
       "45354  ENSG00000175643     RMI2   ENSG00000175643.8"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "differentially_expressed_genes_common_name = convert_ENSEMBL_to_common_name(gtf_df, differentially_expressed_genes)\n",
    "differentially_expressed_genes_common_name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>FBXO45</th>\n",
       "      <th>NCAPG</th>\n",
       "      <th>PTPRB</th>\n",
       "      <th>TRABD2B</th>\n",
       "      <th>H6PD</th>\n",
       "      <th>LCN6</th>\n",
       "      <th>RMI2</th>\n",
       "      <th>RAD54L</th>\n",
       "      <th>ISLR2</th>\n",
       "      <th>Label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>70825.613901</td>\n",
       "      <td>45214.599794</td>\n",
       "      <td>89617.687379</td>\n",
       "      <td>14032.338833</td>\n",
       "      <td>232427.588150</td>\n",
       "      <td>1990.410699</td>\n",
       "      <td>56982.783302</td>\n",
       "      <td>43229.774585</td>\n",
       "      <td>1919.452769</td>\n",
       "      <td>Luminal A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>84313.515188</td>\n",
       "      <td>33888.554894</td>\n",
       "      <td>35178.978153</td>\n",
       "      <td>13246.385298</td>\n",
       "      <td>152172.258997</td>\n",
       "      <td>1412.187531</td>\n",
       "      <td>54930.748805</td>\n",
       "      <td>21248.212379</td>\n",
       "      <td>2237.313823</td>\n",
       "      <td>Luminal A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>98586.080863</td>\n",
       "      <td>27754.410175</td>\n",
       "      <td>41236.180097</td>\n",
       "      <td>21312.785434</td>\n",
       "      <td>310354.839213</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>99935.673890</td>\n",
       "      <td>19625.364375</td>\n",
       "      <td>8941.315316</td>\n",
       "      <td>Luminal A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>97516.942931</td>\n",
       "      <td>45261.276142</td>\n",
       "      <td>39602.273580</td>\n",
       "      <td>9976.604862</td>\n",
       "      <td>181590.072727</td>\n",
       "      <td>9075.270749</td>\n",
       "      <td>253317.484752</td>\n",
       "      <td>46308.038220</td>\n",
       "      <td>9001.787980</td>\n",
       "      <td>Luminal A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>162775.316661</td>\n",
       "      <td>83278.133458</td>\n",
       "      <td>8825.807725</td>\n",
       "      <td>3663.361845</td>\n",
       "      <td>275495.011159</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>431694.167524</td>\n",
       "      <td>67265.099781</td>\n",
       "      <td>1705.678235</td>\n",
       "      <td>Luminal A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>114970.162862</td>\n",
       "      <td>82283.925849</td>\n",
       "      <td>14565.406636</td>\n",
       "      <td>6473.284423</td>\n",
       "      <td>255328.468855</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>160149.679744</td>\n",
       "      <td>64677.674191</td>\n",
       "      <td>5597.412223</td>\n",
       "      <td>Luminal B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>151514.339106</td>\n",
       "      <td>182911.854614</td>\n",
       "      <td>27209.399482</td>\n",
       "      <td>9593.960471</td>\n",
       "      <td>309027.642543</td>\n",
       "      <td>577.627380</td>\n",
       "      <td>135349.258381</td>\n",
       "      <td>64239.016300</td>\n",
       "      <td>4058.398040</td>\n",
       "      <td>Luminal B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>195692.063713</td>\n",
       "      <td>152029.402486</td>\n",
       "      <td>21932.504709</td>\n",
       "      <td>5038.108291</td>\n",
       "      <td>101626.950007</td>\n",
       "      <td>669.254954</td>\n",
       "      <td>320303.042499</td>\n",
       "      <td>125916.506673</td>\n",
       "      <td>7283.755835</td>\n",
       "      <td>Luminal B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>182627.081916</td>\n",
       "      <td>153069.027034</td>\n",
       "      <td>40141.162530</td>\n",
       "      <td>5321.918640</td>\n",
       "      <td>284361.807447</td>\n",
       "      <td>181.049684</td>\n",
       "      <td>139383.253891</td>\n",
       "      <td>85277.106536</td>\n",
       "      <td>1197.224788</td>\n",
       "      <td>Luminal B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>250981.116327</td>\n",
       "      <td>67854.326254</td>\n",
       "      <td>62938.038743</td>\n",
       "      <td>14837.935631</td>\n",
       "      <td>242033.959906</td>\n",
       "      <td>868.364562</td>\n",
       "      <td>89037.025723</td>\n",
       "      <td>25222.448934</td>\n",
       "      <td>10048.889492</td>\n",
       "      <td>Luminal B</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows Ã— 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           FBXO45          NCAPG         PTPRB       TRABD2B           H6PD  \\\n",
       "0    70825.613901   45214.599794  89617.687379  14032.338833  232427.588150   \n",
       "1    84313.515188   33888.554894  35178.978153  13246.385298  152172.258997   \n",
       "2    98586.080863   27754.410175  41236.180097  21312.785434  310354.839213   \n",
       "3    97516.942931   45261.276142  39602.273580   9976.604862  181590.072727   \n",
       "4   162775.316661   83278.133458   8825.807725   3663.361845  275495.011159   \n",
       "..            ...            ...           ...           ...            ...   \n",
       "95  114970.162862   82283.925849  14565.406636   6473.284423  255328.468855   \n",
       "96  151514.339106  182911.854614  27209.399482   9593.960471  309027.642543   \n",
       "97  195692.063713  152029.402486  21932.504709   5038.108291  101626.950007   \n",
       "98  182627.081916  153069.027034  40141.162530   5321.918640  284361.807447   \n",
       "99  250981.116327   67854.326254  62938.038743  14837.935631  242033.959906   \n",
       "\n",
       "           LCN6           RMI2         RAD54L         ISLR2      Label  \n",
       "0   1990.410699   56982.783302   43229.774585   1919.452769  Luminal A  \n",
       "1   1412.187531   54930.748805   21248.212379   2237.313823  Luminal A  \n",
       "2      0.000000   99935.673890   19625.364375   8941.315316  Luminal A  \n",
       "3   9075.270749  253317.484752   46308.038220   9001.787980  Luminal A  \n",
       "4      0.000000  431694.167524   67265.099781   1705.678235  Luminal A  \n",
       "..          ...            ...            ...           ...        ...  \n",
       "95     0.000000  160149.679744   64677.674191   5597.412223  Luminal B  \n",
       "96   577.627380  135349.258381   64239.016300   4058.398040  Luminal B  \n",
       "97   669.254954  320303.042499  125916.506673   7283.755835  Luminal B  \n",
       "98   181.049684  139383.253891   85277.106536   1197.224788  Luminal B  \n",
       "99   868.364562   89037.025723   25222.448934  10048.889492  Luminal B  \n",
       "\n",
       "[100 rows x 10 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "reduced_df = get_reduced_dataframe(luminal_A_df, luminal_B_df, differentially_expressed_genes_common_name, differentially_expressed_genes)\n",
    "reduced_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_gtf(filename, prune = True):\n",
    "    \n",
    "    file = list()\n",
    "    \n",
    "    with open(filename, \"r\") as f:\n",
    "        for line in f: \n",
    "            if line.startswith(\"#\") == False:\n",
    "                row = [x.replace('\\n', '') for x in line.split('\\t')]\n",
    "                \n",
    "                if row[2] == \"gene\":\n",
    "                    row = row[8].split(\";\")\n",
    "                    row = [r for r in row if len(r) > 0]\n",
    "                    \n",
    "                    gene_id   = str(list(filter(lambda x: x != \"\", row[0].split(\"\\\"\")))[1])\n",
    "                    gene_name = str(list(filter(lambda x: x != \"\", row[2].split(\"\\\"\")))[1])\n",
    "                    \n",
    "                    file.append([gene_id, gene_name])\n",
    "        \n",
    "    return file\n",
    "\n",
    "def perform_ttest(luminal_A_df, luminal_B_df, genes):\n",
    "    global BONFERRONI_P_VALUE\n",
    "    \n",
    "    differentially_expressed_genes = list()\n",
    "    \n",
    "    for gene in genes:\n",
    "        t, p = stats.ttest_ind(np.array(luminal_A_df[gene]), np.array(luminal_B_df[gene]), equal_var = False)\n",
    "        \n",
    "        if p < BONFERRONI_P_VALUE:\n",
    "            differentially_expressed_genes.append(gene)\n",
    "            \n",
    "    return {gene.split(\".\")[0]:gene for gene in differentially_expressed_genes}\n",
    "\n",
    "def convert_ENSEMBL_to_common_name(gtf_df, differentially_expressed_genes):\n",
    "   \n",
    "    differentially_expressed_genes_common_name               = gtf_df[gtf_df.GeneId.isin(differentially_expressed_genes.keys())]\n",
    "    differentially_expressed_genes_common_name[\"FullGeneId\"] = differentially_expressed_genes_common_name.GeneId.map(differentially_expressed_genes)\n",
    "    \n",
    "    return differentially_expressed_genes_common_name\n",
    "\n",
    "def get_reduced_dataframe(reduced_luminal_A_df, reduced_luminal_B_df, differentially_expressed_genes_common_name, differentially_expressed_genes):\n",
    "\n",
    "    columns_map = dict(zip(differentially_expressed_genes_common_name.FullGeneId, differentially_expressed_genes_common_name.GeneName)) \n",
    "\n",
    "    reduced_luminal_A_df = luminal_A_df[differentially_expressed_genes.values()].rename(columns = columns_map)\n",
    "    reduced_luminal_B_df = luminal_B_df[differentially_expressed_genes.values()].rename(columns = columns_map)\n",
    "\n",
    "    reduced_luminal_A_df[\"Label\"] = \"Luminal A\"\n",
    "    reduced_luminal_B_df[\"Label\"] = \"Luminal B\"\n",
    "\n",
    "    return pd.concat([reduced_luminal_A_df, reduced_luminal_B_df])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Assignment 2: Use gene expression data to create a classifier for Luminal A / Luminal B breast cancer subtypes\n",
    "\n",
    "In order to create a Luminal A / Luminal B breast cancer classifier, consider two dataset: dataset_LUMINAL_A_B.csv and reduced_dataset.csv.\n",
    "1. Divide both dataset_LUMINAL_A_B.csv and reduced_dateset.csv into train set and test set\n",
    "2. Standardize features of both datasets by removing the mean and scaling to unit variance\n",
    "3. Perform the dimensionality reduction onto dataset.csv with PCA (principal component Analysis) using\n",
    "80 features.\n",
    "4. Train a KNN classifier onto the train set of dataset_LUMINAL_A_B.csv with PCA\n",
    "5. Test the classifier obtained at the previous step onto the test set of dataset_LUMINAL_A_B.csv\n",
    "(remember to apply PCA transformation onto test set)\n",
    "6. Train and test a KNN classifier onto reduced_dataset.csv\n",
    "7. Which are the differences between the performances obtained onto\n",
    "dataset_LUMINAL_A_B.csv with PCA and these obtained onto reduced_dataset.csv?\n",
    "8. Use reduced_dataset.csv to train and test (with default parameters) SVM, Random Forest and\n",
    "NaÃ¯ve Bayes classifiers\n",
    "9. Which classifier provides you the best result on reduced_dataset.csv?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1.\n",
    "luminal_A_B_X_train, luminal_A_B_X_test, luminal_A_B_y_train, luminal_A_B_y_test = split_dataset(luminal_A_B_df)\n",
    "reduced_X_train, reduced_X_test, reduced_y_train, reduced_y_test                 = split_dataset(reduced_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 2.\n",
    "scaler = StandardScaler()\n",
    "\n",
    "scaled_luminal_A_B_X_train, scaled_luminal_A_B_X_test = apply_scaler(scaler, luminal_A_B_X_train, luminal_A_B_X_test)\n",
    "scaled_reduced_X_train, scaled_reduced_X_test         = apply_scaler(scaler, reduced_X_train, reduced_X_test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 3.\n",
    "pca = PCA(n_components=80)\n",
    "\n",
    "dr_luminal_A_B_X_train = pca.fit_transform(luminal_A_B_X_train)\n",
    "dr_luminal_A_B_X_test  = pca.transform    (luminal_A_B_X_test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Accuracy: 0.6'"
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 4, 5.\n",
    "train_classifier(\n",
    "    clf     = KNeighborsClassifier(),\n",
    "    X_train = dr_luminal_A_B_X_train,\n",
    "    y_train = luminal_A_B_y_train,\n",
    "    X_test  = dr_luminal_A_B_X_test,\n",
    "    y_test  = luminal_A_B_y_test\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Accuracy: 0.75'"
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 6.\n",
    "train_classifier(\n",
    "    clf     = KNeighborsClassifier(),\n",
    "    X_train = scaled_reduced_X_train,\n",
    "    y_train = reduced_y_train,\n",
    "    X_test  = scaled_reduced_X_test,\n",
    "    y_test  = reduced_y_test\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Accuracy: 0.9'"
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 8. - Best classifier\n",
    "train_classifier(\n",
    "    clf     = SVC(),\n",
    "    X_train = scaled_reduced_X_train,\n",
    "    y_train = reduced_y_train,\n",
    "    X_test  = scaled_reduced_X_test,\n",
    "    y_test  = reduced_y_test\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Accuracy: 0.75'"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 8.\n",
    "train_classifier(\n",
    "    clf     = RandomForestClassifier(),\n",
    "    X_train = scaled_reduced_X_train,\n",
    "    y_train = reduced_y_train,\n",
    "    X_test  = scaled_reduced_X_test,\n",
    "    y_test  = reduced_y_test\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Accuracy: 0.9'"
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 8. - Best classifier\n",
    "train_classifier(\n",
    "    clf     = GaussianNB(),\n",
    "    X_train = scaled_reduced_X_train,\n",
    "    y_train = reduced_y_train,\n",
    "    X_test  = scaled_reduced_X_test,\n",
    "    y_test  = reduced_y_test\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [],
   "source": [
    "def split_dataset(df):\n",
    "    X = df.drop(\"Label\", axis=1)\n",
    "    y = df[\"Label\"]\n",
    "    return train_test_split(X, y, test_size=0.2)\n",
    "\n",
    "def apply_scaler(scaler, X_train, X_test):\n",
    "    scaled_X_train = scaler.fit_transform(X_train)\n",
    "    scaled_X_test  = scaler.transform    (X_test)\n",
    "    \n",
    "    return scaled_X_train, scaled_X_test\n",
    "\n",
    "def train_classifier(clf, X_train, y_train, X_test, y_test):\n",
    "    \n",
    "    clf.fit(X_train, y_train) \n",
    "    predictions = clf.predict(X_test)\n",
    "\n",
    "    return \"Accuracy: \" + str(accuracy_score(y_test, predictions))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
